services/orchest-webserver/app/app/core/pipelines.py

```
@@ -0,0 +1,162 @@
+        """Remove a pipeline from the FS and the orchest-api"""
+
+        # CleanupPipelineFromOrchest can be used when deleting a
+        # pipeline through Orchest or when cleaning up a pipeline that
+        # was deleted through the filesystrem by the user, so the file
+        # might not be there.
+        with contextlib.suppress(FileNotFoundError):
+            os.remove(self.pipeline_json_path)
+
+        # Orchest-api deletion.
+        url = (
+            f"http://{current_app.config['ORCHEST_API_ADDRESS']}/api/pipelines/"
+            f"{self.project_uuid}/{self.pipeline_uuid}"
+        )
+        current_app.config["SCHEDULER"].add_job(requests.delete, args=[url])
+
+
+class AddPipelineFromFS(TwoPhaseFunction):
+    """Add a pipeline from the FS to Orchest.
+
+    To be used when a pipeline is "discovered" through the FS, e.g. the
+    user has manually added it.
+    """
+
+    def transaction(self, project_uuid: str, pipeline_path: str):
+
+        pipeline_json_path = get_pipeline_path(
+            None, project_uuid, pipeline_path=pipeline_path
+        )
+
+        # Check the uuid of the pipeline. If the uuid is taken by
+        # another pipeline in the project then generate a new uuid for
+        # the pipeline. pipeline_json = json.load(json_file)
+        with open(pipeline_json_path, "r") as json_file:
+            pipeline_json = json.load(json_file)
+            file_pipeline_uuid = pipeline_json.get("uuid")
+            # See if pipeline_uuid is taken or if the pipeline has no
+            # uuid, in both cases it's going to need a new uuid.
+            if (not file_pipeline_uuid) or Pipeline.query.filter(
+                Pipeline.uuid == file_pipeline_uuid
+            ).filter(Pipeline.project_uuid == project_uuid).count() > 0:
+                file_pipeline_uuid = str(uuid.uuid4())
+                # Only write to the file to update the uuid if it's
+                # necessary.
+                self.do_collateral = True
```

I'm seeing this `self.do_collateral` pattern emerge. In the sense that during the transaction one could find out the collateral should not happen. If we were to raise an error the collateral would also not happen (along with bubbling up such that other collaterals part of the TwoPhaseExecutor will not run).

What do you think is preferred? Just not executing this particular collateral? In a way _this_ TwoPhaseFunction does not know in what TwoPhaseExecutor it will be used (by encapsulation). But you could argue that the transaction should raise an exception in the case this particular cannot proceed as intended (giving the callee the responsibility to appropriately handle that).

---

I think we should leave the semantics of what happens in the `transaction` and `collateral` functions to the implementor (excluding commits, ofc). It might be the case that the `transaction` needs to take place while the `collateral` does not need to, I would not consider that an error.

---

As discussed, we'll swap `do_collateral` for passing arguments that can dictate which parts of the collateral function (if any) execute.